include/gridtools/common/int_vector.hpp

```diff
@@ -0,0 +1,134 @@
+/*
+ * GridTools
+ *
+ * Copyright (c) 2014-2021, ETH Zurich
+ * All rights reserved.
+ *
+ * Please, refer to the LICENSE file in the root directory.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/**
+ *  @file
+ * TODO
+ */
+
+#include <type_traits>
+#ifndef GT_TARGET_ITERATING
+//// DON'T USE #pragma once HERE!!!
+#ifndef GT_COMMON_INT_VECTOR_HPP_
+#define GT_COMMON_INT_VECTOR_HPP_
+
+#include "../meta.hpp"
+#include "defs.hpp"
+#include "host_device.hpp"
+#include "hymap.hpp"
+#include "integral_constant.hpp"
+#include "tuple.hpp"
+#include "tuple_util.hpp"
+
+namespace gridtools {
+
+    namespace int_vector_impl_ {
+        template <class... Ts>
+        struct value_t_merger;
+
+        template <class T>
+        struct value_t_merger<T> {
+            using type = T;
+        };
+
+        template <class T0, class T1>
+        struct value_t_merger<T0, T1> {
+            using V0 = meta::second<T0>;
+            using V1 = meta::second<T1>;
+            using K0 = meta::first<T0>;
+            using type = meta::list<K0,
+                meta::if_<integral_constant<bool, is_integral_constant<V0>::value && is_integral_constant<V1>::value>,
+                    decltype(V0{} + V1{}), // TODO this is specific to `plus`
+                    std::common_type_t<V0, V1>>>;
+        };
+
+        template <class... T>
+        using merger_t = typename value_t_merger<T...>::type;
+    } // namespace int_vector_impl_
+
+    namespace int_vector {} // namespace int_vector
+} // namespace gridtools
+
+#define GT_FILENAME <gridtools/common/int_vector.hpp>
+#include GT_ITERATE_ON_TARGETS()
+#undef GT_FILENAME
+
+#endif // GT_COMMON_INT_VECTOR_HPP_
+#else  // GT_TARGET_ITERATING
+
+namespace gridtools {
+    GT_TARGET_NAMESPACE {}
+
+    namespace int_vector {
+        GT_TARGET_NAMESPACE {
+
+            namespace int_vector_detail {
+                template <class I, class Enable = void, class Key = meta::first<I>, class Type = meta::second<I>>
+                struct add_f {
+                    template <class First, class Second>
+                    GT_TARGET GT_FORCE_INLINE GT_TARGET_CONSTEXPR Type operator()(
+                        First &&first, Second &&second) const {
+                        return at_key_with_default<Key, std::integral_constant<Type, 0>>(wstd::forward<First>(first)) +
```

Comment:
how about:
```c++
template <class L, class R>
GT_FUNCTION constexpr auto operator()(L const& l, R const& r) const {
   if constexpr (not has_key<L, Key>())
     return host_device::at_key<Key>(r);
   if constexpr (not has_key<R, Key>())
     return host_device::at_key<Key>(l);
   return host_device::at_key<Key>(l) + host_device::at_key<Key>(r); 
} 
```
Note that there are no forwards here -- args can not be rvalues because they passed to the several generators